worldedit-bukkit/src/main/java/com/sk89q/worldedit/bukkit/adapter/Regenerator.java

```diff
@@ -0,0 +1,559 @@
+import com.sk89q.worldedit.regions.CuboidRegion;
+import com.sk89q.worldedit.regions.Region;
+import com.sk89q.worldedit.world.RegenOptions;
+import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectLinkedOpenHashMap;
+import it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Random;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+import org.bukkit.generator.BlockPopulator;
+
+/**
+ * Represents an abstract regeneration handler.
+ * @param <IChunkAccess> the type of the {@Code IChunkAccess}
+ * @param <ProtoChunk> the type of the {@Code ProtoChunk
+ * @param <Chunk> the type of the {@Code Chunk}
+ * @param <ChunkStatus> the type of the {@Code ChunkStatusWrapper} wrapping the {@Code ChunkStatus} enum
+ */
+public abstract class Regenerator<IChunkAccess, ProtoChunk extends IChunkAccess, Chunk extends IChunkAccess, ChunkStatus extends Regenerator.ChunkStatusWrapper<IChunkAccess>> {
+
+    protected final org.bukkit.World originalBukkitWorld;
+    protected final Region region;
+    protected final Extent target;
+    protected final RegenOptions options;
+
+    //runtime
+    protected final Map<ChunkStatus, Concurrency> chunkStati = new LinkedHashMap<>();
+    protected boolean generateConcurrent = true;
+    protected long seed;
+    
+    private final Long2ObjectLinkedOpenHashMap<ProtoChunk> protoChunks = new Long2ObjectLinkedOpenHashMap<>();
+    private final Long2ObjectOpenHashMap<Chunk> chunks = new Long2ObjectOpenHashMap<>();
+    private ExecutorService executor;
+    private SingleThreadQueueExtent source;
+
+    /**
+     * Initializes an abstract regeneration handler.
+     * @param originalBukkitWorld the Bukkit world containing all the information on how to regenerate the {code Region}
+     * @param region the selection to regenerate
+     * @param target the target {@code Extent} to paste the regenerated blocks into
+     * @param options the options to used while regenerating and pasting into the target {@code Extent}
+     */
+    public Regenerator(org.bukkit.World originalBukkitWorld, Region region, Extent target, RegenOptions options) {
+        this.originalBukkitWorld = originalBukkitWorld;
+        this.region = region;
+        this.target = target;
+        this.options = options;
+    }
+
+    /**
+     * Regenerates the selected {@code Region}.
+     * @return whether or not the regeneration process was successful
+     * @throws Exception when something goes terribly wrong
+     */
+    public boolean regenerate() throws Exception {
+        if (!prepare())
+            return false;
```

Comment:
Everything should be correctly shrouded by {}

Reply:
done